### PR TITLE
Fix undefined behavior in Parquet `varint` decoding

### PR DESCRIPTION
**Pull request by ttnghia**
Fixes undefined behavior in `cudf::io::parquet::detail::CompactProtocolReader::get_varint`, where a shift operation can shift number past the bit width of the result type.

New tests are added to cover such corner cases.

### CODE REVIEW

File: cpp/src/io/parquet/compact_protocol_reader.hpp [hunk 1]
```diff
@@ -51,10 +53,13 @@ class CompactProtocolReader {
   template <typename T>
   T get_varint() noexcept
   {
+    static_assert(std::is_unsigned_v<T>,
+                  "get_varint requires an unsigned T: the shift guard assumes T has no sign bit");
```
Comment (vuule) [inline cpp/src/io/parquet/compact_protocol_reader.hpp]: can this be a `requires`?

File: cpp/src/io/parquet/compact_protocol_reader.hpp
Comment (vuule) [inline cpp/src/io/parquet/compact_protocol_reader.hpp:1]: AFAICT, we should throw when the input is too large, instead of returning a truncated value. `get_varint` is `noexcept`, but it really does not need to be.

**Review by vuule — CHANGES_REQUESTED** [review]

File: cpp/src/io/parquet/compact_protocol_reader.hpp (hunk 1 above)
Comment (igorpeshansky) [reply cpp/src/io/parquet/compact_protocol_reader.hpp]: +1, this would read better as a `requires` than as a `static_assert`. Or we could even use the `std::unsigned_integral` concept.

That said, the only reason for this constraint is the fact that `std::numeric_limits<T>::digits` is too small for signed integers and would drop the top bit for `int64_t`. Otherwise, the signed left-shift overflow is no longer UB as of C++ 20[^1], so signed integers would behave the same as unsigned in that case.

[^1]: https://en.cppreference.com/w/cpp/language/operator_arithmetic#Built-in_bitwise_shift_operators

**Review by igorpeshansky — COMMENTED** [review]
Flagging valid coderabbit complaints (https://github.com/rapidsai/cudf/pull/23346#discussion_r3670506525 and https://github.com/rapidsai/cudf/pull/23346#discussion_r3670506534 )…

File: cpp/src/io/parquet/compact_protocol_reader.hpp [hunk 2]
```diff
@@ -47,13 +53,19 @@ class CompactProtocolReader {
...
     T v = 0;
     for (uint32_t l = 0;; l += 7) {
-      T c = getb();
+      T const c = getb();
+      // Once the shift `l` reaches `T`'s width, shifting this group by `l` is undefined (out of
+      // range): the varint has more groups than `T` can hold, so reject it as overlong/corrupt.
+      CUDF_EXPECTS(l < std::numeric_limits<T>::digits,
+                   "Parquet varint exceeds the width of its target type",
+                   std::overflow_error);
       v |= (c & 0x7f) << l;
```
Comment (wence-) [inline cpp/src/io/parquet/compact_protocol_reader.hpp]: As noted, this check is necessary but not sufficient for correctness.

We need not just that `(1 << l) <= std::numeric_limits<T>::max`, but the stronger condition `((c & 0x7f) << l) <= std::numeric_limits<T>::max`.

So I think you want:

```
CUDF_EXPECTS(l < std::numeric_limits<T>::digits 
             && c <= (std:numeric_limits<T>::max >> l), 
             "Parquet varint exceeds the width of its target type",
             std::overflow_error);
```

You need both conditions because if `l >= digits` then the right shift is undefined.

**Review by wence- — CHANGES_REQUESTED** [review]

File: cpp/src/io/parquet/compact_protocol_reader.hpp (hunk 2 above)
Comment (ttnghia) [reply cpp/src/io/parquet/compact_protocol_reader.hpp]: It seems like we are down in a rabbit hole. I wonder if we have similar snippet like this in the codebase. Will tell my agent to check.

Comment (ttnghia) [reply cpp/src/io/parquet/compact_protocol_reader.hpp]: Confirmed: orc.hpp:414 + avro.cpp:15 (host, same bug); page_hdr.cu:79 (device, no shift guard) + several other device decoders. I'm going to fix this code in this PR, but will defer fixing all others in a separate PR.

Comment (igorpeshansky) [reply cpp/src/io/parquet/compact_protocol_reader.hpp]: Another device instance in cudf-spark-jni: [`spark_rapids_jni::protobuf::detail::read_varint`](https://github.com/NVIDIA/cudf-spark-jni/blob/main/src/main/cpp/src/protobuf/protobuf_device_helpers.cuh#L56).

File: cpp/tests/io/parquet_reader_test.cpp [hunk 3]
```diff
@@ -5574,3 +5575,150 @@ TEST_F(ParquetReaderTest, NestedMismatchedSchemaColumnValidation)
     EXPECT_THROW(cudf::io::read_parquet(opts), std::invalid_argument);
   }
 }
+
+// CompactProtocolReader::get_varint on crafted byte runs, decoded host-side: an overlong varint
```
Comment (mhaseeb123) [inline cpp/tests/io/parquet_reader_test.cpp]: I think these should be moved to a separate file altogether as they are strictly related to the `compact_protocol_reader`

File: cpp/src/io/parquet/compact_protocol_reader.hpp (hunk 2 above)
Comment (ttnghia) [reply cpp/src/io/parquet/compact_protocol_reader.hpp]: Filed issue https://github.com/rapidsai/cudf/issues/23478
For the cudf-spark-jni code, it is not affected since the code around that line already guards against overflow (line 48 and 51 protect from the overflow issue).

File: cpp/tests/io/compact_protocol_reader_test.cpp [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+#include <stdexcept>
+#include <vector>
+
+// CompactProtocolReader::get_varint on crafted byte runs, decoded host-side: an overlong varint
+// (value too large for the target type) throws std::overflow_error; a within-width unterminated run
+// terminates cleanly at EOF; well-formed and empty inputs decode to defined values.
+TEST(CompactProtocolReaderVarintTest, OverlongU32)
+{
+  // A seven-byte run whose value exceeds uint32_t: the value-bound check rejects the fifth byte
+  // 0x85 at shift 28 (0x85 > max() >> 28 == 0x0F), so get_varint throws rather than truncating.
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/tests/io/compact_protocol_reader_test.cpp:22]: False positive. The bound compares the raw byte, not the masked payload.

File: cpp/src/io/parquet/compact_protocol_reader.hpp [hunk 5]
```diff
@@ -36,8 +41,14 @@ class CompactProtocolReader {
...
+    // storage. This keeps every later pointer op defined (the empty state has all-null pointers).
+    CUDF_EXPECTS(base != nullptr || len == 0,
+                 "CompactProtocolReader requires a non-null buffer when length is non-zero",
+                 std::invalid_argument);
     m_base = m_cur = base;
-    m_end          = base + len;
+    // Guard against `nullptr + len` (undefined) so a zero-length buffer stays fully defined.
+    m_end = base != nullptr ? base + len : base;
   }
   [[nodiscard]] ptrdiff_t bytecount() const noexcept { return m_cur - m_base; }
```
Comment (wence-) [inline cpp/src/io/parquet/compact_protocol_reader.hpp]: nit: bytecount is UB if `base` is nullptr.

**Review by wence- — APPROVED** [review]

File: cpp/tests/io/compact_protocol_reader_test.cpp [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+TEST(CompactProtocolReaderVarintTest, NullBufferZeroLengthIsDefinedEmpty)
+{
+  // A null buffer with zero length is valid and yields the fully-defined empty state: get_varint
+  // returns 0 without reading past the null base. Constructed with an explicit nullptr so the
+  // null-base branch of init() is covered regardless of std::vector::data()'s empty-buffer value.
+  cudf::io::parquet::detail::CompactProtocolReader cp(nullptr, 0);
+  EXPECT_EQ(cp.get_u32(), 0u);
+  EXPECT_EQ(cp.bytecount(), static_cast<ptrdiff_t>(0));
+}
```
Comment (mhaseeb123) [inline cpp/tests/io/compact_protocol_reader_test.cpp:161]: Are there any tests here already covered by parquet reader tests and can be removed from here?

**Review by mhaseeb123 — APPROVED** [review]
Can you please post some baseline numbers from the benchmark for posterity

File: cpp/tests/io/compact_protocol_reader_test.cpp (hunk 6 above)
Comment (ttnghia) [reply cpp/tests/io/compact_protocol_reader_test.cpp:161]: Unfortunately not. They call `CompactProtocolReader` directly for which the overflow-throw, EOF, and null/empty-buffer paths are never executed in Parquet reader tests.

**Comment by ttnghia** [comment]
Baseline numbers for the new `PARQUET_VARINT_DECODE_NVBENCH` microbenchmark, for posterity. It is CPU-only (`nvbench::exec_tag::no_gpu`) and times `CompactProtocolReader::get_varint` via `get_u32`/`get_u64` decoding 1,048,576 (2^20) unsigned LEB128 varints per sample, swept by encoded byte width so the per-byte overflow guard is exercised at every width.
```
## `parquet_varint_decode_u32`

| varint_bytes | Samples | CPU Time | Noise | Elem/s |
|--------------|---------|----------|-------|--------|
| 1 | 1248x | 828.519 us | 2.89% | 1.266G |
| 2 | 1440x | 1.050 ms | 4.82% | 999.104M |
| 3 | 2016x | 1.371 ms | 7.75% | 764.678M |
| 4 | 704x | 1.668 ms | 1.88% | 628.532M |
| 5 | 1568x | 1.686 ms | 4.57% | 621.874M |

## `parquet_varint_decode_u64`

| varint_bytes | Samples | CPU Time | Noise | Elem/s |
|--------------|---------|----------|-------|--------|
| 1 | 960x | 838.345 us | 5.06% | 1.251G |
| 2 | 1648x | 833.014 us | 5.32% | 1.259G |
| 3 | 640x | 1.052 ms | 1.99% | 996.345M |
| 4 | 864x | 1.268 ms | 2.54% | 826.942M |
| 5 | 1536x | 1.543 ms | 2.07% | 679.366M |
| 6 | 960x | 1.820 ms | 2.97% | 576.188M |
| 7 | 1136x | 2.202 ms | 3.44% | 476.198M |
| 8 | 240x | 2.433 ms | 2.50% | 431.019M |
| 9 | 1072x | 2.658 ms | 2.89% | 394.449M |
| 10 | 832x | 3.016 ms | 2.31% | 347.691M |
```

Throughput ranges from ~1.27G varints/s at 1 byte down to ~348M/s at 10 bytes, scaling with encoded length as expected.

**Review by igorpeshansky — APPROVED** [review]

File: cpp/tests/io/compact_protocol_reader_test.cpp [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  // 0x85 at shift 28 (0x85 > max() >> 28 == 0x0F), so get_varint throws rather than truncating.
+  std::vector<uint8_t> const bytes{0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x07};
+  cudf::io::parquet::detail::CompactProtocolReader cp(bytes.data(), bytes.size());
+  EXPECT_THROW(cp.get_u32(), std::overflow_error);
+}
+
+TEST(CompactProtocolReaderVarintTest, OverlongU64)
+{
+  // A value exceeding uint64_t: the value-bound check rejects the tenth byte 0x8B at shift 63
+  // (0x8B > max() >> 63 == 0x01), so get_varint throws rather than truncating.
```
Comment (vuule) [inline cpp/tests/io/compact_protocol_reader_test.cpp:31]: Could we remove/shorten some of these comments? I think this one is one of the redundant ones. I can tell from the test name and code what this tests and what it the expected outcome.

File: cpp/benchmarks/io/parquet/parquet_varint_decode.cpp [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+void BM_parquet_varint_decode_u64(nvbench::state& state)
+{
+  bm_parquet_varint_decode<uint64_t>(state);
+}
+
+// u32 varints span 1..5 encoded bytes (32 bits at 7 payload bits per byte, rounded up).
+NVBENCH_BENCH(BM_parquet_varint_decode_u32)
+  .set_name("parquet_varint_decode_u32")
+  .set_is_cpu_only(true)
+  .add_int64_axis("varint_bytes", {1, 2, 3, 4, 5});
```
Comment (vuule) [inline cpp/benchmarks/io/parquet/parquet_varint_decode.cpp:108]:
```suggestion
  .add_int64_axis("varint_bytes", nvbench::range(1, 5));
```

File: cpp/benchmarks/io/parquet/parquet_varint_decode.cpp [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+// u64 varints span 1..10 encoded bytes (64 bits at 7 payload bits per byte, rounded up).
+NVBENCH_BENCH(BM_parquet_varint_decode_u64)
+  .set_name("parquet_varint_decode_u64")
+  .set_is_cpu_only(true)
+  .add_int64_axis("varint_bytes", {1, 2, 3, 4, 5, 6, 7, 8, 9, 10});
```
Comment (vuule) [inline cpp/benchmarks/io/parquet/parquet_varint_decode.cpp:114]:
```suggestion
  .add_int64_axis("varint_bytes", nvbench::range(1, 10));
```

File: cpp/src/io/parquet/compact_protocol_reader.hpp [hunk 10]
```diff
@@ -36,24 +41,43 @@ class CompactProtocolReader {
...
   {
     T v = 0;
     for (uint32_t l = 0;; l += 7) {
-      T c = getb();
+      T const c = getb();
+      // Reject overlong varints. `l < digits` keeps the shift in range (and guards `max() >> l`);
+      // `c <= max() >> l` then bounds the value, since the shift-count check alone would let the
+      // top group's bits silently wrap past `T`'s width instead of being rejected. `c` is the raw
+      // byte (continuation bit included): comparing it, not the masked payload, only rejects a
+      // boundary continuation byte one group early; correct, since the next group overflows anyway.
```
Comment (vuule) [inline cpp/src/io/parquet/compact_protocol_reader.hpp:77]: This could be a bit less verbose.

```suggestion
      // The byte's value, shifted into place, must fit in `T`; `l < digits` also keeps `max() >> l`
      // itself in range. Comparing the raw byte, not the masked payload, is intentional: it also
      // rejects a continuation byte whose successor group could not fit.
```

File: cpp/benchmarks/io/parquet/parquet_varint_decode.cpp
Comment (vuule) [inline cpp/benchmarks/io/parquet/parquet_varint_decode.cpp:1]: Maybe a controversial comment, but do we really need to merge these benchmarks? We verified the impact of the checks. Do we really expect to make future changes to these tiny functions often enough to warrant automated benchmark runs to protect against regressions?

Unless we're planning to tune these functions, this is IMO way too small to need dedicated benchmarks.

**Review by vuule — CHANGES_REQUESTED** [review]
Great change, but altogether my impression is that this should not be a 300 line PR. Left some comments (mostly) in this direction that I would like to at least discuss before merging.

File: cpp/benchmarks/io/parquet/parquet_varint_decode.cpp
Comment (mhaseeb123) [reply cpp/benchmarks/io/parquet/parquet_varint_decode.cpp:1]: Agreed, we don't need to benchmark standalone functions that are rarely used directly.

Comment (mhaseeb123) [reply cpp/benchmarks/io/parquet/parquet_varint_decode.cpp:1]: Same thoughts about tests. We don't need extensive testing of `compact_protocol_reader` per-se.

**Review by mhaseeb123 — CHANGES_REQUESTED** [review]

**Review by wence- — APPROVED** [review]
Approving the core C++ changes with the same misgivings about the benchmarks/test setup as the others.

File: cpp/benchmarks/io/parquet/parquet_varint_decode.cpp
Comment (ttnghia) [reply cpp/benchmarks/io/parquet/parquet_varint_decode.cpp:1]: Removing the benchmarks. However, for tests, all of them are necessary. Each one tests a different encoding vector or code path thus we need to keep all to verify correctness.

File: cpp/tests/io/compact_protocol_reader_test.cpp [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+#include <stdexcept>
+#include <vector>
+
+TEST(CompactProtocolReaderVarintTest, OverlongU32)
+{
+  // Value exceeds uint32_t: the value-bound check rejects byte 0x85 at shift 28.
+  std::vector<uint8_t> const bytes{0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x07};
+  cudf::io::parquet::detail::CompactProtocolReader cp(bytes.data(), bytes.size());
+  EXPECT_THROW(cp.get_u32(), std::overflow_error);
+}
```
Comment (vuule) [inline cpp/tests/io/compact_protocol_reader_test.cpp]: This is redundant with `OverflowAtWidthBoundaryU32` below: both fail the same `c <= max() >> l` check at the same shift (`l == 28`), and the check compares the raw byte, so whether the rejecting byte has the continuation bit set makes no difference to which branch fires. Suggest dropping this test and keeping the boundary one, whose name and comment state the intent more precisely.

File: cpp/tests/io/compact_protocol_reader_test.cpp [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+
+TEST(CompactProtocolReaderVarintTest, OverlongU64)
+{
+  // Value exceeds uint64_t: the value-bound check rejects byte 0x8B at shift 63.
+  std::vector<uint8_t> const bytes{
+    0x81, 0x82, 0x83, 0x84, 0x85, 0x86, 0x87, 0x88, 0x89, 0x8B, 0x8A, 0x00};
+  cudf::io::parquet::detail::CompactProtocolReader cp(bytes.data(), bytes.size());
+  EXPECT_THROW(cp.get_u64(), std::overflow_error);
+}
```
Comment (vuule) [inline cpp/tests/io/compact_protocol_reader_test.cpp]: Same as `OverlongU32`: redundant with `OverflowAtWidthBoundaryU64`, which rejects at the identical shift (`l == 63`) via the identical check. The trailing `0x8A, 0x00` are also never read, since the loop throws on `0x8B`.

File: cpp/tests/io/compact_protocol_reader_test.cpp [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+    std::vector<uint8_t> const bytes{0x80, 0x80, 0x80, 0x80, 0x10};
+    cudf::io::parquet::detail::CompactProtocolReader cp(bytes.data(), bytes.size());
+    EXPECT_THROW(cp.get_i32(), std::overflow_error);
+  }
+  {
+    std::vector<uint8_t> const bytes{0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x02};
+    cudf::io::parquet::detail::CompactProtocolReader cp(bytes.data(), bytes.size());
+    EXPECT_THROW(cp.get_i64(), std::overflow_error);
+  }
+}
```
Comment (vuule) [inline cpp/tests/io/compact_protocol_reader_test.cpp]: Both byte vectors here are byte-for-byte copies of the ones in `OverflowAtWidthBoundaryU32`/`U64`, so the only new coverage is that `get_i32`/`get_i64` propagate rather than swallow the exception on the way through `get_zigzag`. That's worth something now that these are no longer `noexcept`, but one of the two sub-blocks is enough.

File: cpp/tests/io/compact_protocol_reader_test.cpp [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+
+TEST(CompactProtocolReaderVarintTest, OverlongListHeaderThrows)
+{
+  // The size-0xF escape routes through get_u32, which propagates overflow_error.
+  std::vector<uint8_t> const bytes{0xF0, 0x80, 0x80, 0x80, 0x80, 0x10};
+  cudf::io::parquet::detail::CompactProtocolReader cp(bytes.data(), bytes.size());
+  EXPECT_THROW((void)cp.get_listh(), std::overflow_error);
+}
```
Comment (vuule) [inline cpp/tests/io/compact_protocol_reader_test.cpp]: This one is implied by the two tests around it: `WellFormedListHeader` already covers the size-`0xF` escape reaching `get_u32`, and `OverflowAtWidthBoundaryU32` covers `get_u32` throwing on that exact input. `get_listh` has no `noexcept` or catch that could interfere, so I'd drop it.

File: cpp/tests/io/compact_protocol_reader_test.cpp [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+    cudf::io::parquet::detail::CompactProtocolReader cp(bytes.data(), bytes.size());
+    EXPECT_EQ(cp.get_u64(), 0x7'FFFF'FFFFUL);
+    EXPECT_EQ(cp.bytecount(), static_cast<ptrdiff_t>(bytes.size()));
+  }
+  {
+    // However, a run spanning past the width is overlong and throws, terminator or not.
+    std::vector<uint8_t> const bytes(6, 0xFF);
+    cudf::io::parquet::detail::CompactProtocolReader cp(bytes.data(), bytes.size());
+    EXPECT_THROW(cp.get_u32(), std::overflow_error);
+  }
```
Comment (vuule) [inline cpp/tests/io/compact_protocol_reader_test.cpp]: This block is a third copy of the u32 rejection already covered by `OverlongU32` and `OverflowAtWidthBoundaryU32`, and it isn't about the EOF behavior the test is named for -- the run never reaches EOF, it throws on the sixth byte. The first two blocks are the real content of `UnterminatedRunAtEof`; suggest removing this one.

**Review by vuule — CHANGES_REQUESTED** [review]
I think tests can be trimmed down a bit

File: cpp/tests/io/compact_protocol_reader_test.cpp [hunk 16]
```diff
@@ -0,0 +1,165 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <cudf_test/testing_main.hpp>
```
Comment (mhaseeb123) [inline cpp/tests/io/compact_protocol_reader_test.cpp:7]: Can we instead exercise the bug fix test via Parquet write/read path or cut down here as @vuule recommended? Note that we can always add pytests (use pyarrow etc) in pylibcudf or cudf-python if libcudf can't exercise the bug path.

Fine to add the new ones (with Vukasin's comments addressed) if not possible.

Comment (ttnghia) [reply cpp/tests/io/compact_protocol_reader_test.cpp:7]: The overflow inputs cannot be produced through the Parquet write/read path: pyarrow and our writer only write the canonical minimal varints, so an overlong run only exists in hand-crafted data.

**Comment by ttnghia** [comment]
Tests are trimmed down to left the essential ones. Thanks for reviewing.

**Review by vuule — APPROVED** [review]

**Review by mhaseeb123 — APPROVED** [review]

**Comment by mhaseeb123** [comment]
Hi @ttnghia, please rebase and retarget this PR to `release/26.10` branch.

**Comment by ttnghia** [comment]
@coderabbitai re-review.
